libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go

```diff
@@ -769,24 +776,47 @@ func (dhtPeer *DHTPeer) handleNewDelegationConnection(conn net.Conn) {
 			nbrConns.Dec()
 			break
 		}
+		if envel.Sender != addr {
+			err = errors.New("Sender (" + envel.Sender + ") must match registered address")
+			lerror(err).Str("addr", addr).
+				Msg("while routing delegate client envelope")
+			continue
+		}
 
-		// route envelope
-		dhtPeer.goroutines.Add(1)
-		go func() {
-			defer dhtPeer.goroutines.Done()
-			if envel.Sender != addr {
-				err = errors.New("Sender (" + envel.Sender + ") must match registered address")
-				lerror(err).Str("addr", addr).
-					Msg("while routing delegate client envelope")
-			} else {
-				err := dhtPeer.RouteEnvelope(envel)
-				if err != nil {
-					lerror(err).Str("addr", addr).
-						Msg("while routing delegate client envelope")
-					// TODO() send error back
+		// initializing pairwise channel queues and one go routine per pair
+		pair := envel.To + envel.Sender
+		if _, ok := dhtPeer.syncMessages[pair]; !ok {
+			dhtPeer.syncMessages[pair] = make(chan *aea.Envelope, 100)
+
+			// route envelope
+			dhtPeer.goroutines.Add(1)
+			go func() {
+				defer dhtPeer.goroutines.Done()
+
+				for e := range dhtPeer.syncMessages[pair] {
+					fmt.Println("PROCESSING <<<-------- ", string(e.Message))
+					err := dhtPeer.RouteEnvelope(e)
+					fmt.Println("PROCESSED <<<-------- ", string(e.Message))
+					if err != nil {
+						lerror(err).Str("addr", addr).
+							Msg("while routing delegate client envelope")
+						// TODO() send error back
+					}
 				}
-			}
-		}()
+			}()
+		}
+		// add to queue (nonblocking - buffered queue)
```

Comment:
General concept: one producer, multiple consumers (go-routines), one for each pair. This way we ensure we have linerarity for each pair and concurrency across pairs.

Reply:
nice!